src/mongo/db/catalog/collection_options.cpp

```diff
@@ -142,6 +142,8 @@ Status CollectionOptions::parse(const BSONObj& options) {
             cappedSize = e.numberLong();
             if (cappedSize < 0)
                 return Status(ErrorCodes::BadValue, "size has to be >= 0");
+            if ( cappedSize >= 0x7fffffffffffffffLL)
+                return Status(ErrorCodes::BadValue, "size has to be < 2^63");
             cappedSize += 0xff;
```

Comment:
... the problem is really here. This statement causes signed integer overflow for any capped size less than 256 bytes away from the max. The overflow will cause the value to turn negative (or arbitrary badness due to undefined execution) after the check for negative values. The correct fix is to write the alignment expression so it doesn't overflow. Of course checking correct parsing of these huge collections is impractical at the integration level, but correct behavior can be tested in the `collection_options_test` unit test.